### DISCUSSION TITLE
Roadmap, what's next

### DISCUSSION THREAD

**Discussion by lucamonfredo** [EdNekebno/DGTCentaurMods#12] [Ideas]
As per @wormstein request I'm opening this discussion to gather some ideas to move the project forward.

I don't know what is your level of familiarity with Python tools and ecosystem, so I'm just making a random list of things that I'd like to improve, starting from the more general ones:

- definition of a roadmap (even if it's rough, just to update the README with something concrete, see my [pull request](https://github.com/EdNekebno/DGTCentaur/pull/11)). This is just to prioritise things and keep others informed
- move all project files to a dedicated folder (to make it easier to resemble the correct filesystem structure on the Pi, it could be `/home/pi/mods` for example, to go along with `/home/pi/centaur` that will hold the original software)
- modification of all hard coded references in a config file, or something like that, to make it easier to work on the files, as I had to do a couple of edits to make it work on my side. I'm used to put all project config using a [`.env`](https://github.com/theskumar/python-dotenv) file, this way we could easily config everything from a single file
- adding a set up guide in order to have a "standard" and reproducible way of settings things up
- adopt a uniform coding style to make it easier to work together ([Pylint](https://pylint.org))

Regarding the list of features that I would like to implement here are my suggestions:
- PGN export of the games, as a learning player it would be useful for me to be able to collect games and analyse them with an engine afterwards
- Playing Stockfish directly (no Centaur software in between)

These are my thoughts, you have done quite an awesome work with the reverse engineering of the board and I'm really looking forward to contribute and improve the project even more!

Let me know what you think, have a nice day!
Luca

**Comment by wormstein** [comment]
Hey Luca,
I was waiting for project owner input here so that's why I reply so late.

In my opinion this project could be arraged as a python package with modules that handle all board functions such as serial comm, display, menus, sound and then on top of that anyone can expand the functions as they want.

Next, I think a module called something like vboard could be buit and using python chess library it will sit between the board hardware and any chess engine you want. Basically it will make the board a chess GUI. All the board setting could be configured using web interface. Things like upload a new engine, view game history, live board view with real time analyze, etc.
A module to handle a menu would be nice where we could select the type of game (engine or online) that could be easily configured in the web interface.
Other crazy idea: implement socket bluetooth communication and build an app ;)

I am working at the moment on a private repo for some new features but time and knowledge is limited.

There are a ton of cool things that can be developed we just need time :)

-Adrian

**Comment by lucamonfredo** [comment]
Well it's nice that we have so many ideas and try to converge them into one project!

At this point, however, I would suggest limiting the scope and defining some starting points, so as not to waste efforts and time (which is limited for me too) and build a cohesive base to build upon. That's why it would also be interesting the opinion of @EdNekebno that leads the project, in order to start somewhere.

For example, I have already done some work to export the games in PGN format (with the chess lib), I just have to figure out what is the best way to "hook" the generation to some event (start/end of a game, any suggestions maybe?).

Let's see If we can come together with some structure and then prioritise, thank you!
Luca

**Comment by wormstein** [comment]
Did you manage to export the PGN of games of centaur software or the mod?
I think you could use is_check_mate to trigger generation of the PGN.

**Reply by lucamonfredo** [reply]
I extracted those of the centaur but there is no problem in adapting the procedure to everything we decide to do.

**Reply by wormstein** [reply]
If you inject code in centaur, i would just check for check mate and trigger export and send it over email.

Try starting a thread to watch for check mate.

-Adrian

**Comment by EdNekebno** [comment]
Hi, sorry for the slow reply - I was on holiday. Where this is at - I think we now have a  pretty complete idea of the hardware (not fully, maybe it needs a list of missing functionality). You're crazy app idea is not so crazy at all! One of the reasons it's stalled a bit  is that DGT have a new board coming out - https://digitalgametechnology.com/products/home-use-e-boards/dgt-pegasus . This works wireless with an app for android/iOS. Depending on what the app does and how reverse engineerable it is - that might do a lot of the work for us! To me it makes a lot of sense to wait to see what that app does and then use the hardware code on a centaur with the Pi Zero W upgrade to try to emulate that board. (some people are already preparing to do this I think: https://digitalgametechnology.com/products/home-use-e-boards/dgt-pegasus https://github.com/dsommerfeld/DGTCentaur-2.0-by-dso . The same people already have a PGN export: https://schach.computer/erste-erweiterung-fuer-den-dgt-centaur-verfuegbar/ ). Here I'm waiting on the app release and a bit to see what others do as time is short for me.

Everything on the list makes a lot of sense. For my part I'm not a python programmer (or a terribly experienced github user for that matter - it's just a convenient way of making the protocols available), so I'm learning both whilst I'm doing this - hence some very poor coding like not having config files, etc. Definitely feel free to do anything that is tidying up my mess :)

**Comment by lucamonfredo** [comment]
Well it's true, DGT Pegasus promises to have everything we are trying to do here, and maybe more. Despite this, in my opinion it makes sense to try develop at least a subset of what we set out to do, or at least try to ;-)
Let's see what we can do!

**Reply by EdNekebno** [reply]
Oh I agree. I just mean if one of the mods was to make the centaur emulate a pegasus (so you could choose this mode from the menu) then suddenly we'd have a lot of things automatically. Mostly the issue is time and that, if it turns out to be possible, would be reasonably quick.

Most of the ability to play on lichess is done.
Actually on my dev board I have some proxy code (in c I think) sitting between the centaur process and stockfish to grab a pgn - but then it always misses the last move :( But that's adaptable to two of the things you say - it could potentially alter the responses to make the centaur process choose a pure stockfish answer and alter the uci commands to allow different configurations. Also it could read a configuration file somewhere to know what engine to choose.  This, of course, is worth doing regardless of what the pegasus app ultimately does as you can play it purely with the board. Here my issue is just time and that it would almost certainly need an sd card image for people to use and then I need to tidy up mine quite a bit! However, I would really really like a mode that I could beat more than 1 in 4 games :D

**Reply by lucamonfredo** [reply]
Since it seems that you are way ahead of anybody with all the work, would you care to share it with us (maybe in a develop branch) so that we can try to help you for what we can?
Don't worry if it's not complete or functional, it's just to try to not waste time on something you already had done and maybe support you on what's missing.

**Reply by wormstein** [reply]
Stockfish repo in my git repos is modified to pipe current FEN and last UCI command from centaur software. I then experimented a python script to process those pipe files.  You can build the PGN in real time. Just clone, compile and replace the stockfish_pi file in centaur/engines. Bonus you get the new sockfish 14 engine on the board!

**Reply by EdNekebno** [reply]
Nice idea to modify stockfish. I tried to compile the stockfish 14 sources on my board (the originals - not your repo) and couldn't get it to work. Maybe I'll give yours a try.

Proxy that I mention is as follows: rename stockfish_pi (in /home/pi/centaur/engines to stockfish_pi2. Then compile the code and put it in the place of stockfish_pi:

[proxy.txt](https://github.com/EdNekebno/DGTCentaur/files/7218818/proxy.txt)
[uci2pgn.txt](https://github.com/EdNekebno/DGTCentaur/files/7218824/uci2pgn.txt)

For modifications - to choose your own engine for example it would be easiest to put all the engines on the board and then modify the line system("/home/pi/centaur/engines/stockfish_pi2"); to in fact start the engine based on some configuration file. To play pure stockfish or stockfish with specific settings is harder. For the first in theory it needs adapting to modify the return from the engine so that only one possibility is provided or the same possibility multiple times (then the centaur adaptive software only gets one choice - the pure option). For the second it needs to read the uci commands as it comes in and adapt them

Edit: Github discussions mucks up the code formatting - so I've tried attaching them as txt files!

**Reply by wormstein** [reply]
Steps i do

make map
make build ARCH=armv7

Compiled on board straight. There were some headers I had to install. Let me know your error. Maybe I faced that myself too so I can guide you...
